### PR TITLE
HBASE-22853 Git/Jira Release Audit Tool

### PR DESCRIPTION
**Pull request by ndimiduk**
This is an application for performing an audit between the histories on our git branches and the `fixVersion` field set on issues in JIRA. It does this by building a Sqlite database from the commits found on each git branch, identifying Jira IDs and release tags, and then requesting information about those issues from Jira. Once both sources have been collected, queries can be performed against the database to look for discrepancies between the sources of truth (and, possibly, bugs in this script).

### CODE REVIEW

**Comment by ndimiduk** [comment]
Heads up release managers/Jira janitors -- @infraio @Apache9 @busbey @apurtell @saintstack @joshelser

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
||| _ Patch Compile Tests _ |
| -1 :x: |  pylint  |   0m  1s |  Error running pylint. Please check pylint stderr files.  |
| +1 :green_heart: |  pylint  |   0m  1s |  There were no new pylint issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 20s |  The patch does not generate ASF License warnings.  |
|  |   |   2m  6s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1088/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1088 |
| Optional Tests | dupname asflicense markdownlint pylint |
| uname | Linux f59b29dd88c7 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1088/out/precommit/personality/provided.sh |
| git revision | master / eda5df7eac |
| pylint | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1088/1/artifact/out/patch-pylint-stderr.txt |
| Max. process+thread count | 51 (vs. ulimit of 10000) |
| modules | C: . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1088/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) pylint=1.6.5 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by busbey** [comment]
long term, this would make a nice contribution to Apache Yetus.

File: dev-support/git-jira-release-audit/README.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                        branches.
+  --development-branch DEVELOPMENT_BRANCH
+                        The name of the branch from which all release lines
+                        originate.
+  --development-branch-fix-version DEVELOPMENT_BRANCH_FIX_VERSION
+                        The Jira fixVersion used to indicate an issue is
+                        committed to the development branch.
+  --release-line-regexp RELEASE_LINE_REGEXP
+                        A regexp used to identify release lines.
+  --fallback-actions-path FALLBACK_ACTIONS_PATH
```
Comment (busbey) [inline dev-support/git-jira-release-audit/README.md]: Adding a brief section at the end of this README that explains the `fallback-actions-path` is used to identify how the parser should handle commits with known ambiguous messages would help with maintainability 6 months from now.

**Comment by ndimiduk** [comment]
Huh so the usage of pylint is broken in our personality?

**Comment by ndimiduk** [comment]
> long term, this would make a nice contribution to Apache Yetus.

I generally agree, but I wonder how much of this branching style is hbase-specific.

File: dev-support/git-jira-release-audit/README.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+The system prerequisites are Python3 with VirtualEnv available and Sqlite. Also,
+you'll need the content of this directory and a local checkout of git repository.
+
+Build a VirtualEnv with the script's dependencies with
+
+```shell script
+$ python3 --version
+Python 3.7.6
+$ python3 -m venv ./venv
```
Comment (joshelser) [inline dev-support/git-jira-release-audit/README.md]: Is this the "recommended" way of doing it? I've always done: `virtualenv venv && source venv/bin/activate` in the past.

File: dev-support/git-jira-release-audit/README.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  -h, --help            show this help message and exit
+  --db-path DB_PATH     Path to the database file, or leave unspecified for a
+                        transient db.
+  --git-repo-path GIT_REPO_PATH
+                        Path to the git repo, or leave unspecified to infer
+                        from the current file's path.
+  --remote-name REMOTE_NAME
+                        The name of the git remote to use when identifying
+                        branches.
+  --development-branch DEVELOPMENT_BRANCH
```
Comment (joshelser) [inline dev-support/git-jira-release-audit/README.md]: I understand the words you have written, but I'm not sure what I would actually pass to this option. I see in the example below you didn't provide it, but I'm not sure when I would want/need to provide it.

**Comment by busbey** [comment]
> > long term, this would make a nice contribution to Apache Yetus.
> I generally agree, but I wonder how much of this branching style is hbase-specific.

yeah, I'm totally on board with us using it locally here first and then seeing if it's suitable for others later.

> Huh so the usage of pylint is broken in our personality?

shoot, is it? I wonder how long ago that happened. can someone get up a jira with a description and I'll try to find someone to chase it down?

File: dev-support/git-jira-release-audit/README.md (hunk 1 above)
Comment (ndimiduk) [reply dev-support/git-jira-release-audit/README.md]: I've added a comment to the top of that file.

**Comment by ndimiduk** [comment]
Address PR feedback, burn down the pylint complaints.

**Comment by ndimiduk** [comment]
> shoot, is it? I wonder how long ago that happened. can someone get up a jira with a description and I'll try to find someone to chase it down?

https://issues.apache.org/jira/browse/HBASE-23745

File: dev-support/git-jira-release-audit/README.md (hunk 2 above)
Comment (ndimiduk) [reply dev-support/git-jira-release-audit/README.md]: Dunno what you mean by "recommended". This is python -- there are multiple ways to solve a problem. In this case, I wanted to be sure I was getting virtualenv from a python3 install, hence the `-mvenv` invocation. I also have no need to "drop into" the environment, so `venv/bin/activate` doesn't buy me anything.

File: dev-support/git-jira-release-audit/README.md (hunk 3 above)
Comment (ndimiduk) [reply dev-support/git-jira-release-audit/README.md]: So for HBase and 95% of git-tracked projects, this is `master`. I believe Hadoop calls the `trunk`, as per its Subversion legacy. It's the branch where development takes place; I'm not sure what else to call this.

File: dev-support/git-jira-release-audit/README.md (hunk 2 above)
Comment (joshelser) [reply dev-support/git-jira-release-audit/README.md]: Oh, virtualenv used to be a completely separate thing and it's been rolled into python 3.3 under the `venv` module (https://virtualenv.pypa.io/en/latest/). I guess this approach eliminates getting some "wrong" virtualenv executable. Just curious, not an issue :)

File: dev-support/git-jira-release-audit/README.md (hunk 3 above)
Comment (joshelser) [reply dev-support/git-jira-release-audit/README.md]: Gotcha. Wasn't sure if there was a case where we'd be providing a value for hbase. I understand the hadoop use-case. Thanks!

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
||| _ Patch Compile Tests _ |
| -1 :x: |  pylint  |   0m  0s |  Error running pylint. Please check pylint stderr files.  |
| +1 :green_heart: |  pylint  |   0m  0s |  There were no new pylint issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 18s |  The patch does not generate ASF License warnings.  |
|  |   |   1m 51s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1088/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1088 |
| Optional Tests | dupname asflicense markdownlint pylint |
| uname | Linux abb235115cfb 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1088/out/precommit/personality/provided.sh |
| git revision | master / 375e01dac4 |
| pylint | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1088/2/artifact/out/patch-pylint-stderr.txt |
| Max. process+thread count | 51 (vs. ulimit of 10000) |
| modules | C: . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1088/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) pylint=1.6.5 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**